distribution/std-lib/Database/src/Connection/Connection.enso

```diff
@@ -23,7 +35,182 @@ type Connection
 
        Arguments:
        - name: name of the table to access
-    access_table : Text -> Table
-    access_table name =
-        column_names = this.internal_connection.fetch_column_names name
-        Table.make_table this name column_names
+    access_table : Text -> Database_Table
+    access_table name = here.wrap_sql_errors <|
+        columns = this.fetch_columns name
+        Database_Table.make_table this name columns
+
+    ## Closes the connection releasing the underlying database resources
+       immediately instead of waiting for them to be automatically released.
+
+       The connection is not usable afterwards.
+    close : Nothing
+    close =
+        Managed_Resource.finalize this.connection_resource
+
+    ## ADVANCED
+
+       Executes a raw query and returns the result as an in-memory Table.
+
+       Arguments:
+       - query: either raw SQL code as Text or an instance of Sql.Statement
+         representing the query to execute.
+       - expected_types: an optional array of expected types of each column;
+         meant only for internal use.
+    execute_query : Text | Sql.Statement -> Vector Sql.Sql_Type -> Materialized_Table =
+    execute_query query expected_types=Nothing = here.wrap_sql_errors <|
+        Resource.bracket (this.prepare_statement query) .close stmt->
+            rs = stmt.executeQuery
+            metadata = rs.getMetaData
+            ncols = metadata.getColumnCount
+            ## TODO [RW] we wrap everything in objects, in the future we may
```

Comment:
The thing is, I only know the types more-or-less - the result could for example include a VARCHAR or a TEXT or LONGTEXT or one of hundeds of other SQL types from some obscure engine, I use `getObject` which handles this inference automatically for me (with the exception of boolean) and otherwise I'd need to create some kind of type mapping that would be engine specific.

Reply:
Ok so to sum up our follow-up discussion, I try to detect at least the standard SQL types and use the more optimized builders for them, otherwise I fallback to the default InferredBuilder, but I now use it directly instead of instantiating a temporary Vector, so even this fallback codepath should be pretty fast.